feat(apple): add manual file I/O tracing and experimental swizzling options (#12999)
getsentry/sentry-docs: docs/platforms/apple/common/tracing/instrumentation/automatic-instrumentation.mdx

}];
```

> Starting with iOS 18, macOS 15 and tvOS 18, the `NSFileManager` is not backed by `NSData` anymore and needs to be swizzled directly.

Reviewer: ```suggestion
Starting with iOS 18, macOS 15, and tvOS 18, the `NSFileManager` is no longer backed by `NSData` and needs to be swizzled directly.
```

> This feature is currently experimental and disabled by default, as it might cause issues with the internal behaviour of these classes.

Author: Makes sense, do you think we should mention why it is disabled by default?

> This feature is currently experimental and disabled by default, as it might cause issues with the internal behaviour of these classes.

Reviewer 2: Yes please.

> This feature is currently experimental and disabled by default, as it might cause issues with the internal behaviour of these classes.

Author: What would be a better phrasing for the reason of it being disabled by default because swizzling is a sensitive topic?

> This feature is currently experimental and disabled by default, as it might cause issues with the internal behaviour of these classes.

Reviewer 2: It's experimental, that's why 😅.

To use automatic file I/O tracing for `NSFileManager` you need to enable it:

```swift {tabTitle:Swift}
import Sentry

SentrySDK.start { options in
    options.dsn = "___PUBLIC_DSN___"
    options.enableFileIOTracing = true
    options.experimental.enableFileManagerSwizzling = true
}
```

```objc {tabTitle:Objective-C}
@import Sentry;

[SentrySDK startWithConfigureOptions:^(SentryOptions *options) {
    options.dsn = "___PUBLIC_DSN___"
    options.enableFileIOTracing = YES;
    options.experimental.enableFileManagerSwizzling = YES;
}];
```

> Starting with iOS 18, tvOS 18 and macOS 15 the Swift classes `Data` and `FileManager` are not backed by their Objective-C counterparts `NSData` and `NSFileManager` anymore and can therefore not be instrumented automatically anymore. Instead, please use [Manual File I/O Tracing](#manual-fileio-tracing).

Reviewer: ```suggestion
Starting with iOS 18, tvOS 18, and macOS 15 the Swift classes `Data` and `FileManager` are no longer backed by their Objective-C counterparts `NSData` and `NSFileManager` and therefore cannot be instrumented automatically anymore. Instead, please use [Manual File I/O Tracing](#manual-fileio-tracing).
```

[NSData]: https://developer.apple.com/documentation/foundation/nsdata
[NSFileManager]: https://developer.apple.com/documentation/foundation/nsfilemanager
[NSString]: https://developer.apple.com/documentation/foundation/nsstring
[NSBundle]: https://developer.apple.com/documentation/foundation/nsbundle

### Manual File/IO Tracing

> The Sentry SDK offers extensions for the types `Data` and `FileManager` to track reading and writing data, as well as creating, moving, copying and deleting files.

Reviewer: ```suggestion
The Sentry SDK offers extensions for the types `Data` and `FileManager` to track reading and writing data, as well as creating, moving, copying, and deleting files.
```
